python/cudf/cudf/core/buffer/spill_manager.py

```diff
@@ -151,6 +155,30 @@ def log_expose(self, buf: SpillableBuffer) -> None:
                 stat.total_nbytes += buf.nbytes
                 stat.spilled_nbytes += spilled_nbytes
 
+    def log_spill_on_demand(self, manager: SpillManager, nbytes: int):
+        """Log an spill-on-demand event (after it has been handled)
+
+        Parameters
+        ----------
+        manager : SpillManager
+            The manager evoking the event
+        nbytes : int
+            Number of bytes requested by the event
+        """
+
+        if self.level < 3:
+            return
+
+        unspilled = sum(
+            buf.size for buf in manager.buffers() if not buf.is_spilled
+        )
+        if self.lowest_utilization is None:
+            self.lowest_utilization = unspilled + nbytes
+        else:
+            self.lowest_utilization = min(
+                self.lowest_utilization, unspilled + nbytes
+            )
```

Comment:
You are both absolutely right. I have to think some more about this.  

I used the current implementation to investigate strange spilling behavior. It worked because most allocations were similar sized and not too large (~100MB). In some extreme cases, spilling-on-demand for a 100MB allocation would trigger at a device utilization of 25GB (on a 32GB GPU). This let me to the idea of tracking the device utilization but as you both points out, this is not a good measurement. 